kernel/src/process.rs

```diff
@@ -1143,43 +1290,190 @@ impl<C: Chip> ProcessType for Process<'_, C> {
+            // Clippy complains that we're deferencing a pointer in a
+            // public and safe function here. While we are not
+            // deferencing the pointer here, we pass it along to an
+            // unsafe function, which is as dangerous (as it is likely
+            // to be deferenced down the line).
+            //
+            // Relevant discussion:
+            // https://github.com/rust-lang/rust-clippy/issues/3045
+            //
+            // It should be fine to ignore the lint here, as a slice
+            // of length 0 will never allow dereferencing any memory
+            // in a safe manner.
+            unsafe { ReadWriteAppSlice::new(buf_start_addr, 0, self.appid()) }
+        } else if self.in_app_owned_memory(buf_start_addr, size) {
+            // TODO: Check for buffer aliasing here
+
+            // Valid slice, we need to adjust the app's watermark
+            // note: in_app_owned_memory ensures this offset does not wrap
+            let buf_end_addr = buf_start_addr.wrapping_add(size);
+            let new_water_mark = max(self.allow_high_water_mark.get(), buf_end_addr);
+            self.allow_high_water_mark.set(new_water_mark);
+
+            // Clippy complains that we're deferencing a pointer in a
+            // public and safe function here. While we are not
+            // deferencing the pointer here, we pass it along to an
+            // unsafe function, which is as dangerous (as it is likely
+            // to be deferenced down the line).
+            //
+            // Relevant discussion:
+            // https://github.com/rust-lang/rust-clippy/issues/3045
+            //
+            // It should be fine to ignore the lint here, as long as
+            // we make sure that we're pointing towards userspace
+            // memory (verified using `in_app_owned_memory`) and
+            // respect alignment and other constraints of the Rust
+            // references created by ReadWriteAppSlice.
+            unsafe { ReadWriteAppSlice::new(buf_start_addr, size, self.appid()) }
+        } else {
+            return GenericSyscallReturnValue::AllowReadWriteFailure(
+                ErrorCode::INVAL,
+                buf_start_addr,
+                size,
+            );
+        };
+
+        let allow_result = driver_invoc_closure(new_slice);
+
+        // TODO: Check whether the capsule either made the allow fail
+        // and returned the new slice, or it made the allow succeed
+        // and returned the expected slice from the allow table
+
+        match allow_result {
+            Ok(old_slice) => {
+                let (ptr, len) = old_slice.consume();
+                GenericSyscallReturnValue::AllowReadWriteSuccess(ptr, len)
+            }
+            Err((new_slice, err)) => {
+                let (ptr, len) = new_slice.consume();
+                GenericSyscallReturnValue::AllowReadWriteFailure(err, ptr, len)
+            }
+        }
```

Comment:
It's still not quite clear to me why this logic is handled here in process.rs. It seems like all process.rs needs to do is create an appslice or not (as the 1.x API was). I think handling the semantics of the syscall is what sched.rs does.

Reply:
I'll defer to @lschuermann on this.

Reply:
@bradjc I agree, handling system call semantics is the primary job of the scheduler. However it's pretty clear that -- given the scheduler's `handle_syscall` method's size and complexity already -- we're going to want to use a separate function / method for this.

My primary motivation to make this a method on `Process` vs. having a static function in `sched.rs` was that AppSlices and Callbacks are a process resource, and the Tock 2.0 system call ABI emphasizes this further. Especially the checks for Callback and AppSlice swap prevention utilize the fact that the resources of each process can be managed separately, in the respective Grant region. Furthermore, the AppSlice table will be stored in the process' memory region and thus implemented (at least in part) in `process.rs`. Also, issues such as #2369 show us that enforcing the invariants of slices required by Rust are difficult to keep track of in such a complex software. Having everything related to creating references into process memory inside `process.rs` and enforce the invariants there seemed like the sane option.

Note that this method in `process.rs` is _only_ responsible for resource allocation (e.g. creating the `AppSlice`). _If_ that is granted (all invariants are checked and validated), a closure in `sched.rs` is invoked which does the actual system call routing to the driver, so `sched.rs` is still responsible for that.

Hope this makes sense. We could probably move it to `sched.rs` as well, but maybe these arguments might convince you that it's a good idea to keep it this way.

Reply:
`handle_syscall` is a major part of the kernel, so I would expect it to be complex. I actually think the code there is pretty clear, but moving the logic on how syscalls are handled to multiple files makes it harder to understand exactly how syscalls are handled.

AppSlices are memory regions in processes, sure, but their semantics and management are separate from process management, and they exist in a separate file. Callbacks contain a pointer in the process memory space, but they too are separate from process management. process.rs can verify that app slices and callbacks are actually within a process, but how they are used is outside of the scope of process.rs.

I think that deciding what return value to send back to a process as result of a system call executed on a Driver is part of the syscall ABI, and should be handled in sched.rs. Process is complex and subtle, and I'm happy to offload complexity :). Also, the use of a closure adds complexity, and isn't used elsewhere in the kernel, so it makes the code a little more difficult to read (i.e. why does allow() use a closure to process.rs, but not, say memop()?).

Reply:
I get the point of saying that's it's not a good idea to build the actual return value returned to the process in process.rs, given that, as you said, it's part of the ABI and should be of concern to the scheduler and architecture, but not the process.

You've managed to convince me to move it to `sched.rs`, but I'd appreciate if we can get this molded into at least a separate function there, reducing the cognitive load posed by the (maybe easy to understand but still) giant `match` struct here. Hopefully the integration of the allow-table to prevent slice aliasing won't decrease cohesion / increase coupling between those two files even more.

I still feel like neither is a really great solution, looking at #2369 and #2464 I think it's obvious that we need to be really careful about how, where and when we reference process memory, and the tight integration with the scheduler required here doesn't exactly help with that.

Reply:
I don't have any issue with a new function.

Reply:
I've moved the code in question back to `sched.rs` in #2487. It's better for now, we'll see how the integration looks once the allow-buffer aliasing prevention table mechanism is implemented, but this can be reviewed later & separately.

For what it's worth, on `rv32i` the closure cause a net change of 0 in the instruction count for handling an allow system call on `rv32i` compared to what I propose in #2487.